coresimd/ppsv/api/shuffles.rs

```diff
@@ -0,0 +1,391 @@
+                )
+            };
+        r
+        }
+    };
+    ($vec0:expr, $vec1:expr,
+     [$l0:expr, $l1:expr, $l2:expr, $l3:expr,
+      $l4:expr, $l5:expr, $l6:expr, $l7:expr,
+      $l8:expr, $l9:expr, $l10:expr, $l11:expr,
+      $l12:expr, $l13:expr, $l14:expr, $l15:expr,
+      $l16:expr, $l17:expr, $l18:expr, $l19:expr,
+      $l20:expr, $l21:expr, $l22:expr, $l23:expr,
+      $l24:expr, $l25:expr, $l26:expr, $l27:expr,
+      $l28:expr, $l29:expr, $l30:expr, $l31:expr]) => {
+        {
+            #[allow(unused_unsafe)]
+            let r = unsafe {
+                $crate::simd::__shuffle_vector32(
+                    $vec0, $vec1,
+                    [$l0, $l1, $l2, $l3,
+                     $l4, $l5, $l6, $l7,
+                     $l8, $l9, $l10, $l11,
+                     $l12, $l13, $l14, $l15,
+                     $l16, $l17, $l18, $l19,
+                     $l20, $l21, $l22, $l23,
+                     $l24, $l25, $l26, $l27,
+                     $l28, $l29, $l30, $l31]
+                )
+            };
+            r
+        }
+    };
+    ($vec0:expr, $vec1:expr,
+     [$l0:expr, $l1:expr, $l2:expr, $l3:expr,
+      $l4:expr, $l5:expr, $l6:expr, $l7:expr,
+      $l8:expr, $l9:expr, $l10:expr, $l11:expr,
+      $l12:expr, $l13:expr, $l14:expr, $l15:expr,
+      $l16:expr, $l17:expr, $l18:expr, $l19:expr,
+      $l20:expr, $l21:expr, $l22:expr, $l23:expr,
+      $l24:expr, $l25:expr, $l26:expr, $l27:expr,
+      $l28:expr, $l29:expr, $l30:expr, $l31:expr,
+      $l32:expr, $l33:expr, $l34:expr, $l35:expr,
+      $l36:expr, $l37:expr, $l38:expr, $l39:expr,
+      $l40:expr, $l41:expr, $l42:expr, $l43:expr,
+      $l44:expr, $l45:expr, $l46:expr, $l47:expr,
+      $l48:expr, $l49:expr, $l50:expr, $l51:expr,
+      $l52:expr, $l53:expr, $l54:expr, $l55:expr,
+      $l56:expr, $l57:expr, $l58:expr, $l59:expr,
+      $l60:expr, $l61:expr, $l62:expr, $l63:expr]) => {
+        {
+            #[allow(unused_unsafe)]
+            let r = unsafe {
+                $crate::simd::__shuffle_vector64(
+                    $vec0, $vec1,
+                    [$l0, $l1, $l2, $l3,
+                     $l4, $l5, $l6, $l7,
+                     $l8, $l9, $l10, $l11,
+                     $l12, $l13, $l14, $l15,
+                     $l16, $l17, $l18, $l19,
+                     $l20, $l21, $l22, $l23,
+                     $l24, $l25, $l26, $l27,
+                     $l28, $l29, $l30, $l31,
+                     $l32, $l33, $l34, $l35,
+                     $l36, $l37, $l38, $l39,
+                     $l40, $l41, $l42, $l43,
+                     $l44, $l45, $l46, $l47,
+                     $l48, $l49, $l50, $l51,
+                     $l52, $l53, $l54, $l55,
+                     $l56, $l57, $l58, $l59,
+                     $l60, $l61, $l62, $l63]
+                )
+            };
+            r
+        }
+    };
+    ($vec:expr, [$($l:expr),*]) => {
+        shuffle!($vec, $vec, [$($l),*])
```

Comment:
This will evaluate `$vec` twice.

Reply:
@Amanieu 

What's the best way to fix this? Just:

```rust
{
    let v = $vec;
    shuffle!(v, v, ...)
}
```

?

Reply:
`match $vec { v => shuffle!(v, v, ...) }` is preferred because of `let ... in ...` semantics, which regular `let` doesn't have (less relevant here, but all the temporaries in `$vec` stay alive for the duration of the `match`).